### ISSUE TITLE
Machine Learning Algorithm -- Nearest Neighbor Methods

### ISSUE THREAD

**Issue by deactivated-account-by-me** [FabianTianyiNi/articles#4]
### basic theory
**Learning:** Store all the training examples, like <a href="https://www.codecogs.com/eqnedit.php?latex=(x^{(i)},&space;y^{(i)})" target="_blank"><img src="https://latex.codecogs.com/gif.latex?(x^{(i)},&space;y^{(i)})" title="(x^{(i)}, y^{(i)})" /></a>. Within it, the <img src="https://latex.codecogs.com/gif.latex?x^{(i)}" title="x^{(i)}" /> is the ordinates of the trainning points while the <img src="https://latex.codecogs.com/gif.latex?y^{(i)}" title="y^{(i)}" /> is the label property.
**Classification:** The **_first_** step is to find the training example <a href="https://www.codecogs.com/eqnedit.php?latex=(x^{(i)},&space;y^{(i)})" target="_blank"><img src="https://latex.codecogs.com/gif.latex?(x^{(i)},&space;y^{(i)})" title="(x^{(i)}, y^{(i)})" /></a>, such that the <img src="https://latex.codecogs.com/gif.latex?x^{(i)}" title="x^{(i)}" /> is the closest point to the new point <img src="https://latex.codecogs.com/gif.latex?x{}'" title="x{}'" />. The **_second_** step is to classify the new point <img src="https://latex.codecogs.com/gif.latex?x{}'" title="x{}'" /> with the label of <img src="https://latex.codecogs.com/gif.latex?x^{(i)}" title="x^{(i)}" />, that is <img src="https://latex.codecogs.com/gif.latex?y^{(i)}" title="y^{(i)}" />.

### Challenges
* How to choose the right measure of closeness?  That is how to measure and find the closest point of the new point?

**Euclidean distance** is popular, but many other possibilities.
As we know, the Euclidean distance is between the point p and the point q, which is a line segment connecting them. The distance between p and q is as follows,

<img src="https://wikimedia.org/api/rest_v1/media/math/render/svg/dc0281a964ec758cca02ab9ef91a7f54ac00d4b7"/>

However, the Euclidean distance makes sense only if when each of the features is roughly on the 
 **same scale**, or the coordinates of the points cannot be put in the equation to calculate. 
To better solve the problem above, the feature vectors are often scaled by the standard deviation over the training set.
#### classification -- K-Dimensional Trees
We use k-d trees to help simplify the classification task by constructing a tree that partitions the search space.

1. Starting with the entire training set, choose some dimension, _i_.
1. Select an element of the training data whose _ith_ dimension has the median value among all elements of the training set.
1. Divide the training set into two pieces: depending on if their _ith_ attribute is smaller or larger than the median.
1. Repeat this partitioning process on each of the two new pieces separately.

**_example:_**
We assume in the training data, there are 7 points.
Based on dimension 1(X):
(1, 1) (2, 9.5) (3, 5) (4, 2) (5, 9) (7, 5.5) (8, 4)
Based on dimension 2(Y):
(1, 1) (4, 2) (8, 4) (3, 5) (7, 5.5) (5, 9) (2, 9.5)

* How to choose k
-- Too small k and the estimates are noisy, while too large and the accuracy suffers

* What if the nearest neighbor is really far away?
